src/v/bytes/details/io_fragment.h

```diff
@@ -90,6 +90,7 @@ class io_fragment {
     void trim(size_t len) { _used_bytes = std::min(len, _used_bytes); }
     void trim_front(size_t pos) {
         // required by input_stream<char> converter
+        _used_bytes -= std::min(pos, _buf.size());
```

Comment:
I'm not sure this is correct. Part of the confusion might be the naming of `_used_bytes` which by all accounts--based on its name--should be decremented after trimming the front.

But it looks like used_bytes in this class moves forward only. For example, it seems to point at the end of the usable space (see `get_current` below), and i'm not sure that should change if we trim the front.

This may still be a bug, but I'm not sure yet. Still thinking. Good too to have alex take a look.

Reply:
I think that `get_current` gives you the pointer to write between the `size()` and `capacity()`. The `size()` returns `_used_bytes`. Also, from the first glance it looks that byte_iterator and iterator_consumer wont be working correctly if 'trim_front' was called. I can try to reproduce.

Reply:
@Lazin  - w.r.t iterator invalidation, can you write a test. We have effectively a unique pointer to the fragment, so iterators should be stable

Reply:
also, @Lazin  maybe what we should do is `_buf.trim_front` should actually return the bytes trimmed or throw 

and we decrease bytes used by that

Reply:
i think it doesn't make sense that the caller (iobuf)would call this method with *more* bytes to trim from the front.

Reply:
You're right, the check is not necessary. I removed it. Also, added commit that reproduces the error with make_input_stream and io_iterator_consumer/io_byte_iterator.